File: src/lib/events.ts
/**
 * The Rust→React event channel — typed subscriptions over the generated
 * `events` object (tauri-specta). Event names AND payload types come from
 * the same generated definition the Rust side emits through, so neither
 * side can drift; this layer only adds the lifecycle convention:
 * `onX(cb)` returns a synchronous unsubscribe, so effect cleanups and
 * try/finally blocks never juggle unlisten promises.
 *
 * The help-window protocol has no generated form — `emitTo` (main window
 * → help webview) is not part of tauri-specta's generated surface — so
 * its names and payloads live here too: this is the single module that
 * knows them. `HELP_WINDOW_LABEL`/`HelpTarget` are defined here and
 * re-exported from help-window.ts for existing importers.
 */
import { emitTo, listen } from '@tauri-apps/api/event'
import { events } from '@/lib/tauri-bindings'
import type {
  SessionSnapshot,
  SetlistExportProgressEvent,
  WindowStateSnapshot,
} from '@/lib/tauri-bindings'
import type { ColorTheme } from '@/lib/color-theme'

/** The help webview's Tauri window label. */
export const HELP_WINDOW_LABEL = 'help'

/** Where the help window should navigate: the search box or one document. */
export type HelpTarget = { kind: 'search' } | { kind: 'doc'; docId: string }

/** Synchronous unsubscribe — the unlisten promise resolves internally. */
export type Unsubscribe = () => void

function toUnsubscribe(pending: Promise<() => void>): Unsubscribe {
  return () => {
    void pending.then(off => off())
  }
}

/** Every session snapshot publication (the state machine's funnel). */
export function onSessionSnapshot(
  cb: (snapshot: SessionSnapshot) => void
): Unsubscribe {
  return toUnsubscribe(
    events.sessionSnapshotEvent.listen(e => cb(e.payload.snapshot))
  )
}

/** Window chrome state (fullscreen, traffic lights, fade generation). */
export function onWindowState(
  cb: (snapshot: WindowStateSnapshot) => void
): Unsubscribe {
  return toUnsubscribe(
    events.windowStateEvent.listen(e => cb(e.payload.snapshot))
  )
}

/**
 * The window regained macOS focus (emitted payload-less; consumers
 * re-derive from stores — AppShell restores the session snapshot,
 * MonitorView reclaims the keyboard focus).
 */
export function onWindowFocus(cb: () => void): Unsubscribe {
  return toUnsubscribe(events.windowFocusEvent.listen(() => cb()))
}

/** A queued `.pnds` bundle finished installing — open it. */
export function onOpenBundle(cb: () => void): Unsubscribe {
  return toUnsubscribe(events.openBundleEvent.listen(() => cb()))
}

/** Setlist folder export progress, once per packed project. */
export function onSetlistExportProgress(
  cb: (progress: SetlistExportProgressEvent) => void
): Unsubscribe {
  return toUnsubscribe(
    events.setlistExportProgressEvent.listen(e => cb(e.payload))
  )
}

// ---------------------------------------------------------------------------
// Help window protocol (emitTo-targeted; names held only here)
// ---------------------------------------------------------------------------

/** The help webview finished booting and is listening for navigation. */
export function onHelpReady(cb: () => void): Unsubscribe {
  return toUnsubscribe(events.helpReadyEvent.listen(() => cb()))
}

/** HelpCenter side of the ready handshake. */
export function emitHelpReady(): Promise<void> {
  return events.helpReadyEvent.emit({})
}

const HELP_NAVIGATE = 'pnds:help-navigate'
const HELP_LOCALE = 'pnds:help-locale'
const HELP_THEME = 'pnds:help-theme'

/** HelpCenter side: receive navigation pushed from the main window. */
export function onHelpNavigate(cb: (target: HelpTarget) => void): Unsubscribe {
  return toUnsubscribe(listen<HelpTarget>(HELP_NAVIGATE, e => cb(e.payload)))
}

/** Main-window side: navigate the (booted) help window. */
export function emitHelpNavigate(target: HelpTarget): Promise<void> {
  return emitTo(HELP_WINDOW_LABEL, HELP_NAVIGATE, target)
}

/** HelpCenter side: follow the main window's UI language. */
export function onHelpLocale(cb: (locale: string) => void): Unsubscribe {
  return toUnsubscribe(
    listen<{ locale: string }>(HELP_LOCALE, e => cb(e.payload.locale))
  )
}

/** Main-window side: push a UI-language change to the help window. */
export function emitHelpLocale(locale: string): Promise<void> {
  return emitTo(HELP_WINDOW_LABEL, HELP_LOCALE, { locale })
}

/** HelpCenter side: follow the main window's color theme. */
export function onHelpTheme(cb: (theme: ColorTheme) => void): Unsubscribe {
  return toUnsubscribe(
    listen<{ colorTheme: ColorTheme }>(HELP_THEME, e =>
      cb(e.payload.colorTheme)
    )
  )
}

/** Main-window side: push a color-theme change to the help window. */
export function emitHelpTheme(theme: ColorTheme): Promise<void> {
  return emitTo(HELP_WINDOW_LABEL, HELP_THEME, { colorTheme: theme })
}

File: src-tauri/src/events.rs
//! App-internal Tauri events — the typed Rust→React channel.
//!
//! Each event is a payload struct registered in `collect_events!`
//! (bindings.rs); tauri-specta derives the wire name from the struct's
//! name and generates the frontend's typed `events` object from the same
//! definition, so neither side can drift. The reverse channel (React →
//! Rust) is the generated `commands`; this module is the other half.
//!
//! Naming history: these were stringly `"pnds:*"` events with hand-typed
//! payloads at every listener. The names are App-internal (both sides
//! regenerate atomically) and moved to the derive's struct names in one
//! pass — no external contract referenced the old strings.

use serde::{Deserialize, Serialize};
use specta::Type;
use tauri_specta::Event;

use crate::project::session::SessionSnapshot;
use crate::window::WindowStateSnapshot;

/// Session state publication — every snapshot the state machine emits
/// (formerly `pnds:session`); statuses per runtime-contract §8/§9.
#[derive(Debug, Clone, Serialize, Deserialize, Type, Event)]
pub struct SessionSnapshotEvent {
    pub snapshot: SessionSnapshot,
}

/// Window chrome state publication (formerly `pnds:window`).
#[derive(Debug, Clone, Serialize, Deserialize, Type, Event)]
pub struct WindowStateEvent {
    pub snapshot: WindowStateSnapshot,
}

/// The window regained macOS focus (formerly `pnds:window-focus`).
/// Payload-less by design: every consumer re-derives from the stores
/// (AppShell restores the session snapshot, MonitorView reclaims the
/// keyboard focus). lib.rs re-fires it on a delay schedule because an
/// emit into a still-suspended webview is dropped outright.
#[derive(Debug, Clone, Serialize, Deserialize, Type, Event)]
pub struct WindowFocusEvent {}

/// A queued `.pnds` bundle finished installing (formerly
/// `pnds:open-bundle`) — the frontend opens the installed project.
#[derive(Debug, Clone, Serialize, Deserialize, Type, Event)]
pub struct OpenBundleEvent {}

/// Setlist folder export progress (formerly
/// `pnds:setlist-export-progress`), emitted once per packed project.
#[derive(Debug, Clone, Serialize, Deserialize, Type, Event)]
#[serde(rename_all = "camelCase")]
pub struct SetlistExportProgressEvent {
    // u32 (not usize) so the generated TS type is number, not BigInt —
    // specta's default export config forbids BigInt.
    pub done: u32,
    pub total: u32,
    pub file_name: String,
}

/// The help window finished booting and is listening for navigation
/// (formerly `pnds:help-ready`). Emitted by the help webview itself;
/// registered here so both windows share one generated name.
#[derive(Debug, Clone, Serialize, Deserialize, Type, Event)]
pub struct HelpReadyEvent {}

/// The App's event set as a fresh builder for any runtime. The real app
/// composes it with commands (bindings.rs); tests mount it on
/// mock-runtime apps so `Event::emit` resolves its registry entry.
pub fn events_builder<R: tauri::Runtime>() -> tauri_specta::Builder<R> {
    tauri_specta::Builder::<R>::new().events(tauri_specta::collect_events![
        SessionSnapshotEvent,
        WindowStateEvent,
        WindowFocusEvent,
        OpenBundleEvent,
        SetlistExportProgressEvent,
        HelpReadyEvent,
    ])
}
